velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/prestosql/types/JsonType.h"
+
+#include "folly/Conv.h"
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/functions/lib/StringEncodingUtils.h"
+
+namespace facebook::velox {
+
+bool JsonCastOperator::isSupportedType(const TypePtr& other) const {
+  switch (other->kind()) {
+    case TypeKind::BOOLEAN:
+    case TypeKind::BIGINT:
+    case TypeKind::INTEGER:
+    case TypeKind::SMALLINT:
+    case TypeKind::TINYINT:
+    case TypeKind::DOUBLE:
+    case TypeKind::REAL:
+    case TypeKind::VARCHAR:
+    case TypeKind::DATE:
+    case TypeKind::TIMESTAMP:
+      return true;
+    case TypeKind::ARRAY:
+      return isSupportedType(other->childAt(0));
+    case TypeKind::ROW:
+      for (auto& child : other->as<TypeKind::ROW>().children()) {
+        if (!isSupportedType(child)) {
+          return false;
+        }
+      }
+      return true;
+    case TypeKind::MAP:
+      return (
+          other->childAt(0)->kind() == TypeKind::VARCHAR &&
+          isSupportedType(other->childAt(1)));
+    default:
+      return false;
+  }
+}
+
+template <typename T>
+std::string generateJsonTyped(const SimpleVector<T>& inputBase, int row) {
+  std::string result;
+
+  auto value = inputBase.valueAt(row);
+  folly::toAppend<std::string, T>(value, &result);
+
+  return result;
+}
+
+void JsonCastOperator::castTo(
+    const DecodedVector& inputDecoded,
+    const SelectivityVector& baseRows,
+    const TypePtr& fromType,
+    VectorPtr& result) {
+  // localResult is ensured writable before passed in.
+  auto* flatResult = result->as<FlatVector<StringView>>();
+
+  if (fromType->isPrimitiveType()) {
+    baseRows.applyToSelected([&](auto row) {
+      std::string jsonString;
+
+      switch (fromType->kind()) {
```

Comment:
Can we move that switch outside of the loop? The value is the same for all rows.

Reply:
Switched the order. Sorry for missing this in this previous version.